File: lstm/scala/lstm.scala
class LSTMCell(val units: Int) {
  private val kernel = Array(4)(Array(units))
  private val bias = Array(4)(Float)

  def apply(inputs: Array[Float], h_t_prev: Array[Float], c_t_prev: Array[Float], h_t: Array[Float], c_t: Array[Float]): Unit = {
    // Apply the input gate
    val i_t = sigmoid(kernel(0) dot inputs + bias(0))

    // Apply the forget gate
    val f_t = sigmoid(kernel(1) dot inputs + bias(1))

    // Apply the cell gate
    val c_t_hat = tanh(kernel(2) dot inputs + bias(2))

    // Apply the output gate
    val o_t = sigmoid(kernel(3) dot inputs + bias(3))

    // Update the cell state
    c_t = f_t * c_t_prev + i_t * c_t_hat

    // Output the hidden state
    h_t = o_t * tanh(c_t)
  }
}

object Main {
  def main(args: Array[String]): Unit = {
    // Initialize the LSTM cell
    val cell = LSTMCell(10)

    // Create the input and output sequences
    val inputs = Array(0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0)
    val outputs = Array(0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0)

    // Initialize the hidden state and cell state
    val h_t_prev = Array(0.0)
    val c_t_prev = Array(0.0)

    // Apply the LSTM cell for each time step
    for (t <- 0 until 10) {
      // Apply the LSTM cell
      cell.apply(inputs.slice(t, None), h_t_prev, c_t_prev, h_t, c_t)

      // Update the hidden state and cell state
      h_t_prev = h_t
      c_t_prev = c_t
    }

    // Print the output sequence
    println(h_t.mkString(", "))
  }
}

File: lstm/ruby/lstm.rb
class LSTMCell
    attr_reader :units
  
    def initialize(units)
      @units = units
      @kernel = [
        [1.0, 2.0, 3.0, 4.0],
        [10.0, 20.0, 30.0, 40.0],
        [100.0, 200.0, 300.0, 400.0],
        [1000.0, 2000.0, 3000.0, 4000.0],
      ]
      @bias = [1.0, 2.0, 3.0, 4.0]
    end
  
    def apply(inputs, h_t_prev, c_t_prev, h_t, c_t)
      # Apply the input gate
      i_t = sigmoid(inputs.dot(@kernel[0]) + @bias[0])
  
      # Apply the forget gate
      f_t = sigmoid(inputs.dot(@kernel[1]) + @bias[1])
  
      # Apply the cell gate
      c_t_hat = tanh(inputs.dot(@kernel[2]) + @bias[2])
  
      # Apply the output gate
      o_t = sigmoid(inputs.dot(@kernel[3]) + @bias[3])
  
      # Update the cell state
      c_t = f_t * c_t_prev + i_t * c_t_hat
  
      # Output the hidden state
      h_t = o_t * tanh(c_t)
    end
  end
  
  def main
    # Initialize the LSTM cell
    cell = LSTMCell(10)
  
    # Create the input and output sequences
    inputs = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
    outputs = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
  
    # Initialize the hidden state and cell state
    h_t_prev = [0.0]
    c_t_prev = [0.0]
  
    # Apply the LSTM cell for each time step
    for t in 0..9
      # Apply the LSTM cell
      cell.apply(inputs[t:], h_t_prev, c_t_prev, h_t, c_t)
  
      # Update the hidden state and cell state
      h_t_prev = h_t
      c_t_prev = c_t
    end
  
    # Print the output sequence
    puts h_t
  end
  
  main
  